File: packages/nesd_audio/lib/nesd_audio.dart
/// Push-model PCM audio output for NESd.
///
/// Thin FFI wrapper around the native `nesd_audio` library. All
/// methods must be called from the isolate that opened the stream;
/// the OS audio callback communicates with it only through the
/// native SPSC ring.
library;

import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

typedef _OpenNative = Pointer<Void> Function(Int32, Int32, Int32, Int32, Int32);
typedef _Open = Pointer<Void> Function(int, int, int, int, int);

typedef _PushNative = Int32 Function(Pointer<Void>, Pointer<Float>, Int32);
typedef _Push = int Function(Pointer<Void>, Pointer<Float>, int);

typedef _VoidFnNative = Void Function(Pointer<Void>);
typedef _VoidFn = void Function(Pointer<Void>);

typedef _Int32FnNative = Int32 Function(Pointer<Void>);
typedef _Int32Fn = int Function(Pointer<Void>);

typedef _Uint32FnNative = Uint32 Function(Pointer<Void>);

/// Kind of device backing a [NesdAudio] stream.
enum NesdAudioState {
  /// Playing on a real output device.
  realDevice,

  /// The null device was requested via [NesdAudio.open] `nullDevice`.
  nullDevice,

  /// No real device was available; fell back to the null device.
  nullFallback,
}

class NesdAudio {
  NesdAudio._(this._handle, this._bindings);

  /// Opens a playback stream of f32 samples.
  ///
  /// [bufferSamples] is the native ring capacity; after an underrun
  /// the stream emits silence until the ring refills to
  /// [recoverSamples]. [nullDevice] selects a timer-driven fake
  /// device that consumes samples in real time without touching
  /// audio hardware.
  factory NesdAudio.open({
    required int sampleRate,
    required int channels,
    required int bufferSamples,
    required int recoverSamples,
    bool nullDevice = false,
  }) {
    final bindings = _Bindings(_load());

    final handle = bindings.open(
      sampleRate,
      channels,
      bufferSamples,
      recoverSamples,
      nullDevice ? _flagNullDevice : 0,
    );

    if (handle == nullptr) {
      throw StateError('Failed to open the nesd_audio stream');
    }

    return NesdAudio._(handle, bindings);
  }

  /// Overrides where the native library is loaded from.
  ///
  /// A per-isolate static, like `Lz4Codec.libraryPath`: tests set it
  /// in `flutter_test_config.dart` and forward it to worker isolates
  /// via `NesIsolateConfig`. Null in production builds (the plugin
  /// bundles the library).
  static String? libraryPath;

  static const _flagNullDevice = 1;

  final Pointer<Void> _handle;
  final _Bindings _bindings;

  Pointer<Float> _pushBuffer = nullptr;
  int _pushBufferCapacity = 0;

  /// Ring capacity in samples.
  int get capacity => _bindings.capacity(_handle);

  /// Samples currently buffered in the native ring.
  int get filled => _bindings.filled(_handle);

  /// Underrun count since the last [resetStats].
  int get underruns => _bindings.underruns(_handle);

  /// Overrun (short write) count since the last [resetStats].
  int get overruns => _bindings.overruns(_handle);

  /// Lifetime count of device restarts after OS-initiated stops.
  int get restarts => _bindings.restarts(_handle);

  NesdAudioState get state => NesdAudioState.values[_bindings.state(_handle)];

  /// Pushes samples into the native ring; returns how many were
  /// written. A short write means the ring was near capacity.
  int push(Float32List samples) {
    if (samples.isEmpty) {
      return 0;
    }

    if (_pushBufferCapacity < samples.length) {
      if (_pushBuffer != nullptr) {
        calloc.free(_pushBuffer);
      }

      _pushBuffer = calloc<Float>(samples.length);
      _pushBufferCapacity = samples.length;
    }

    _pushBuffer.asTypedList(samples.length).setAll(0, samples);

    return _bindings.push(_handle, _pushBuffer, samples.length);
  }

  void resetStats() => _bindings.resetStats(_handle);

  void close() {
    _bindings.close(_handle);

    if (_pushBuffer != nullptr) {
      calloc.free(_pushBuffer);
      _pushBuffer = nullptr;
      _pushBufferCapacity = 0;
    }
  }

  static DynamicLibrary _load() {
    if (libraryPath case final path?) {
      return DynamicLibrary.open(path);
    }

    if (Platform.isWindows) {
      return DynamicLibrary.open('nesd_audio.dll');
    }

    if (Platform.isMacOS) {
      return DynamicLibrary.open('nesd_audio.framework/nesd_audio');
    }

    return DynamicLibrary.open('libnesd_audio.so');
  }
}

class _Bindings {
  _Bindings(DynamicLibrary library)
    : open = library
          .lookup<NativeFunction<_OpenNative>>('nesd_audio_open')
          .asFunction(),
      close = library
          .lookup<NativeFunction<_VoidFnNative>>('nesd_audio_close')
          .asFunction(),
      push = library
          .lookup<NativeFunction<_PushNative>>('nesd_audio_push')
          .asFunction(),
      capacity = library
          .lookup<NativeFunction<_Int32FnNative>>('nesd_audio_capacity')
          .asFunction(),
      filled = library
          .lookup<NativeFunction<_Int32FnNative>>('nesd_audio_filled')
          .asFunction(),
      state = library
          .lookup<NativeFunction<_Int32FnNative>>('nesd_audio_state')
          .asFunction(),
      underruns = library
          .lookup<NativeFunction<_Uint32FnNative>>('nesd_audio_underruns')
          .asFunction(),
      overruns = library
          .lookup<NativeFunction<_Uint32FnNative>>('nesd_audio_overruns')
          .asFunction(),
      restarts = library
          .lookup<NativeFunction<_Uint32FnNative>>('nesd_audio_restarts')
          .asFunction(),
      resetStats = library
          .lookup<NativeFunction<_VoidFnNative>>('nesd_audio_reset_stats')
          .asFunction();

  final _Open open;
  final _VoidFn close;
  final _Push push;
  final _Int32Fn capacity;
  final _Int32Fn filled;
  final _Int32Fn state;
  final _Int32Fn underruns;
  final _Int32Fn overruns;
  final _Int32Fn restarts;
  final _VoidFn resetStats;
}

File: packages/nesd_audio/src/nesd_audio.cc
#include "nesd_audio.h"

#include <atomic>
#include <cstring>
#include <new>

#include "nesd_audio_backend.h"
#include "spsc_ring.h"

struct nesd_audio {
    spsc_ring_t ring;
    int32_t recover_samples;
    int32_t state;

    bool is_exhaust;          // consumer-thread state only
    uint32_t pop_watermark;   // consumer-thread state only

    std::atomic<uint32_t> underruns;
    std::atomic<uint32_t> overruns;

    uint32_t restarts;  // producer-thread mirror of the backend count

    nesd_audio_backend_t *backend;
};

// Runs on the OS audio thread. After an underrun, keeps emitting
// silence until the ring holds the largest single pop the device has
// made plus recover_samples of margin, so playback can survive the next
// pop with margin. Resuming below one pop would hand the next pop a ring
// it drains straight back to empty, cascading underruns.
static void render(void *user, float *out, int32_t samples) {
    auto *stream = (nesd_audio_t *)user;

    if ((uint32_t)samples > stream->pop_watermark) {
        stream->pop_watermark = (uint32_t)samples;
    }

    int32_t recover =
        stream->recover_samples + (int32_t)stream->pop_watermark;

    // clamp recovery target to capacity so it is reachable
    if (recover > (int32_t)stream->ring.logical_size) {
        recover = (int32_t)stream->ring.logical_size;
    }

    if (stream->is_exhaust &&
        (int32_t)spsc_filled(&stream->ring) < recover) {
        memset(out, 0, (size_t)samples * sizeof(float));
        stream->underruns.fetch_add(1, std::memory_order_relaxed);
        return;
    }

    stream->is_exhaust = false;

    uint32_t copied = spsc_pop(&stream->ring, out, (uint32_t)samples);

    if ((int32_t)copied < samples) {
        memset(out + copied, 0,
               ((size_t)samples - copied) * sizeof(float));
        stream->is_exhaust = true;
        stream->underruns.fetch_add(1, std::memory_order_relaxed);
    }
}

nesd_audio_t *nesd_audio_open(int32_t sample_rate, int32_t channels,
                              int32_t buffer_samples,
                              int32_t recover_samples,
                              int32_t flags) {
    auto *stream = new (std::nothrow) nesd_audio_t();

    if (stream == NULL) {
        return NULL;
    }

    spsc_init(&stream->ring, (uint32_t)buffer_samples);
    stream->recover_samples = recover_samples;
    stream->is_exhaust = false;
    stream->pop_watermark = 0;
    stream->underruns.store(0, std::memory_order_relaxed);
    stream->overruns.store(0, std::memory_order_relaxed);
    stream->restarts = 0;

    const bool want_null =
        (flags & NESD_AUDIO_FLAG_NULL_DEVICE) != 0;

    stream->backend = nesd_audio_backend_open(
        sample_rate, channels, want_null, render, stream);

    if (stream->backend == NULL && !want_null) {
        stream->backend = nesd_audio_backend_open(
            sample_rate, channels, true, render, stream);
    }

    if (stream->backend == NULL) {
        spsc_destroy(&stream->ring);
        delete stream;
        return NULL;
    }

    if (want_null) {
        stream->state = NESD_AUDIO_STATE_NULL_DEVICE;
    } else if (nesd_audio_backend_is_null_device(stream->backend)) {
        stream->state = NESD_AUDIO_STATE_NULL_FALLBACK;
    } else {
        stream->state = NESD_AUDIO_STATE_REAL_DEVICE;
    }

    return stream;
}

void nesd_audio_close(nesd_audio_t *stream) {
    nesd_audio_backend_close(stream->backend);
    spsc_destroy(&stream->ring);
    delete stream;
}

int32_t nesd_audio_push(nesd_audio_t *stream, const float *samples,
                        int32_t count) {
    stream->restarts =
        nesd_audio_backend_ensure_running(stream->backend);

    int32_t free_samples = (int32_t)stream->ring.logical_size -
                           (int32_t)spsc_filled(&stream->ring);
    int32_t writable = count < free_samples ? count : free_samples;

    if (writable < count) {
        stream->overruns.fetch_add(1, std::memory_order_relaxed);
    }

    if (writable <= 0) {
        return 0;
    }

    spsc_push(&stream->ring, samples, (uint32_t)writable);

    return writable;
}

int32_t nesd_audio_capacity(nesd_audio_t *stream) {
    return (int32_t)stream->ring.logical_size;
}

int32_t nesd_audio_filled(nesd_audio_t *stream) {
    return (int32_t)spsc_filled(&stream->ring);
}

int32_t nesd_audio_state(nesd_audio_t *stream) {
    return stream->state;
}

uint32_t nesd_audio_underruns(nesd_audio_t *stream) {
    return stream->underruns.load(std::memory_order_relaxed);
}

uint32_t nesd_audio_overruns(nesd_audio_t *stream) {
    return stream->overruns.load(std::memory_order_relaxed);
}

uint32_t nesd_audio_restarts(nesd_audio_t *stream) {
    return stream->restarts;
}

void nesd_audio_reset_stats(nesd_audio_t *stream) {
    stream->underruns.store(0, std::memory_order_relaxed);
    stream->overruns.store(0, std::memory_order_relaxed);
}
